arm64 Linux prebuilts

Could prebuilts be added for arm64 Linux? I'm wasting around ten minutes in ci right now waiting for this to build each time.

happy to help with configuring ci and whatever to accomplish this

I can't get it to successfully build on arm64 linux (AWS Graviton).

```
167 error gyp info it worked if it ends with ok
167 error gyp info using node-gyp@10.0.1
167 error gyp info using node@16.15.0 | linux | arm64
167 error gyp info find Python using Python version 3.7.10 found at "/bin/python3"
167 error gyp WARN EACCES current user ("webapp") does not have permission to access the dev dir "/root/.cache/node-gyp/16.15.0"
167 error gyp WARN EACCES attempting to reinstall using temporary dev dir "/tmp/.node-gyp"
167 error gyp info spawn /bin/python3
167 error gyp info spawn args [
167 error gyp info spawn args '/var/app/staging/node_modules/node-gyp/gyp/gyp_main.py',
167 error gyp info spawn args 'binding.gyp',
167 error gyp info spawn args '-f',
167 error gyp info spawn args 'make',
167 error gyp info spawn args '-I',
167 error gyp info spawn args '/var/app/staging/node_modules/re2/build/config.gypi',
167 error gyp info spawn args '-I',
167 error gyp info spawn args '/var/app/staging/node_modules/node-gyp/addon.gypi',
167 error gyp info spawn args '-I',
167 error gyp info spawn args '/tmp/.node-gyp/16.15.0/include/node/common.gypi',
167 error gyp info spawn args '-Dlibrary=shared_library',
167 error gyp info spawn args '-Dvisibility=default',
167 error gyp info spawn args '-Dnode_root_dir=/tmp/.node-gyp/16.15.0',
167 error gyp info spawn args '-Dnode_gyp_dir=/var/app/staging/node_modules/node-gyp',
167 error gyp info spawn args '-Dnode_lib_file=/tmp/.node-gyp/16.15.0/<(target_arch)/node.lib',
167 error gyp info spawn args '-Dmodule_root_dir=/var/app/staging/node_modules/re2',
167 error gyp info spawn args '-Dnode_engine=v8',
167 error gyp info spawn args '--depth=.',
167 error gyp info spawn args '--no-parallel',
167 error gyp info spawn args '--generator-output',
167 error gyp info spawn args 'build',
167 error gyp info spawn args '-Goutput_dir=.'
167 error gyp info spawn args ]
167 error gyp info spawn make
167 error gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
167 error g++: error: unrecognized command line option '-std=c++2a'; did you mean '-std=c++03'?
167 error make: *** [Release/obj.target/re2/lib/addon.o] Error 1
```

@AAGaming00 did you have to do anything special (like install some os deps 1st) to get it to build via gyp?

> I can't get it to successfully build on arm64 linux (AWS Graviton).
> 
> ```
> 167 error gyp info it worked if it ends with ok
> 167 error gyp info using node-gyp@10.0.1
> 167 error gyp info using node@16.15.0 | linux | arm64
> 167 error gyp info find Python using Python version 3.7.10 found at "/bin/python3"
> 167 error gyp WARN EACCES current user ("webapp") does not have permission to access the dev dir "/root/.cache/node-gyp/16.15.0"
> 167 error gyp WARN EACCES attempting to reinstall using temporary dev dir "/tmp/.node-gyp"
> 167 error gyp info spawn /bin/python3
> 167 error gyp info spawn args [
> 167 error gyp info spawn args '/var/app/staging/node_modules/node-gyp/gyp/gyp_main.py',
> 167 error gyp info spawn args 'binding.gyp',
> 167 error gyp info spawn args '-f',
> 167 error gyp info spawn args 'make',
> 167 error gyp info spawn args '-I',
> 167 error gyp info spawn args '/var/app/staging/node_modules/re2/build/config.gypi',
> 167 error gyp info spawn args '-I',
> 167 error gyp info spawn args '/var/app/staging/node_modules/node-gyp/addon.gypi',
> 167 error gyp info spawn args '-I',
> 167 error gyp info spawn args '/tmp/.node-gyp/16.15.0/include/node/common.gypi',
> 167 error gyp info spawn args '-Dlibrary=shared_library',
> 167 error gyp info spawn args '-Dvisibility=default',
> 167 error gyp info spawn args '-Dnode_root_dir=/tmp/.node-gyp/16.15.0',
> 167 error gyp info spawn args '-Dnode_gyp_dir=/var/app/staging/node_modules/node-gyp',
> 167 error gyp info spawn args '-Dnode_lib_file=/tmp/.node-gyp/16.15.0/<(target_arch)/node.lib',
> 167 error gyp info spawn args '-Dmodule_root_dir=/var/app/staging/node_modules/re2',
> 167 error gyp info spawn args '-Dnode_engine=v8',
> 167 error gyp info spawn args '--depth=.',
> 167 error gyp info spawn args '--no-parallel',
> 167 error gyp info spawn args '--generator-output',
> 167 error gyp info spawn args 'build',
> 167 error gyp info spawn args '-Goutput_dir=.'
> 167 error gyp info spawn args ]
> 167 error gyp info spawn make
> 167 error gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
> 167 error g++: error: unrecognized command line option '-std=c++2a'; did you mean '-std=c++03'?
> 167 error make: *** [Release/obj.target/re2/lib/addon.o] Error 1
> ```
> 
> @AAGaming00 did you have to do anything special (like install some os deps 1st) to get it to build via gyp?

distro issue, compiler too old but it seems like you already realized that

also set JOBS=max if you don't want the build to take forever

--- a/.github/actions/linux-arm64-node-18/Dockerfile
+++ b/.github/actions/linux-arm64-node-18/Dockerfile
@@ -1,0 +1,6 @@
+FROM --platform=linux/arm64 node:18-buster
+
+RUN apt install python3 make gcc g++
+
+COPY entrypoint.sh /entrypoint.sh
+ENTRYPOINT ["/entrypoint.sh"]

--- a/.github/actions/linux-arm64-node-18/action.yml
+++ b/.github/actions/linux-arm64-node-18/action.yml
@@ -1,0 +1,12 @@
+name: 'Create a binary artifact for Node == 18 on Linux on ARM64'
+description: 'Create a binary artifact for Node == 18 on Linux on ARM64 using node:18-buster'
+inputs:
+  node-version:
+    description: 'Node.js version'
+    required: false
+    default: '18'
+runs:
+  using: 'docker'
+  image: 'Dockerfile'
+  args:
+    - ${{inputs.node-version}}

--- a/.github/actions/linux-arm64-node-18/entrypoint.sh
+++ b/.github/actions/linux-arm64-node-18/entrypoint.sh
@@ -1,0 +1,10 @@
+#!/bin/sh
+
+set -e
+
+export USERNAME=`whoami`
+export DEVELOPMENT_SKIP_GETTING_ASSET=true
+npm i
+npm run build --if-present
+npm test
+npm run save-to-github

--- a/.github/actions/linux-arm64-node-20/Dockerfile
+++ b/.github/actions/linux-arm64-node-20/Dockerfile
@@ -1,0 +1,6 @@
+FROM --platform=linux/arm64 node:20-buster
+
+RUN apt install python3 make gcc g++
+
+COPY entrypoint.sh /entrypoint.sh
+ENTRYPOINT ["/entrypoint.sh"]

--- a/.github/actions/linux-arm64-node-20/action.yml
+++ b/.github/actions/linux-arm64-node-20/action.yml
@@ -1,0 +1,12 @@
+name: 'Create a binary artifact for Node 20 on Linux on ARM64'
+description: 'Create a binary artifact for Node 20 on Linux on ARM64 using node:20-buster'
+inputs:
+  node-version:
+    description: 'Node.js version'
+    required: false
+    default: '20'
+runs:
+  using: 'docker'
+  image: 'Dockerfile'
+  args:
+    - ${{inputs.node-version}}

--- a/.github/actions/linux-arm64-node-20/entrypoint.sh
+++ b/.github/actions/linux-arm64-node-20/entrypoint.sh
@@ -1,0 +1,10 @@
+#!/bin/sh
+
+set -e
+
+export USERNAME=`whoami`
+export DEVELOPMENT_SKIP_GETTING_ASSET=true
+npm i
+npm run build --if-present
+npm test
+npm run save-to-github

--- a/.github/actions/linux-arm64-node-21/Dockerfile
+++ b/.github/actions/linux-arm64-node-21/Dockerfile
@@ -1,0 +1,6 @@
+FROM --platform=linux/arm64 node:21-bullseye
+
+RUN apt install python3 make gcc g++
+
+COPY entrypoint.sh /entrypoint.sh
+ENTRYPOINT ["/entrypoint.sh"]

--- a/.github/actions/linux-arm64-node-21/action.yml
+++ b/.github/actions/linux-arm64-node-21/action.yml
@@ -1,0 +1,12 @@
+name: 'Create a binary artifact for Node 21 on Linux on ARM64'
+description: 'Create a binary artifact for Node 21 on Linux on ARM64 using node:21-bullseye'
+inputs:
+  node-version:
+    description: 'Node.js version'
+    required: false
+    default: '21'
+runs:
+  using: 'docker'
+  image: 'Dockerfile'
+  args:
+    - ${{inputs.node-version}}

--- a/.github/actions/linux-arm64-node-21/entrypoint.sh
+++ b/.github/actions/linux-arm64-node-21/entrypoint.sh
@@ -1,0 +1,10 @@
+#!/bin/sh
+
+set -e
+
+export USERNAME=`whoami`
+export DEVELOPMENT_SKIP_GETTING_ASSET=true
+npm i
+npm run build --if-present
+npm test
+npm run save-to-github

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -159,3 +159,41 @@ jobs:
       uses: ./.github/actions/linux-arm64-node-18/
       env:
         GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+
+  build-linux-arm64-node-20:
+    name: Node.js 20 on Debian Buster on ARM64
+    needs: create-release
+    runs-on: ubuntu-latest
+    continue-on-error: true
+
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        submodules: true
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v3
+      with:
+        platforms: arm64
+    - name: Install, test, and create artifact
+      uses: ./.github/actions/linux-arm64-node-20/
+      env:
+        GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+
+  build-linux-arm64-node-21:
+    name: Node.js 21 on Debian Buster on ARM64
+    needs: create-release
+    runs-on: ubuntu-latest
+    continue-on-error: true
+
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        submodules: true
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v3
+      with:
+        platforms: arm64
+    - name: Install, test, and create artifact
+      uses: ./.github/actions/linux-arm64-node-21/
+      env:
+        GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -140,3 +140,22 @@ jobs:
       uses: ./.github/actions/linux-alpine-node-21/
       env:
         GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+
+  build-linux-arm64-node-18:
+    name: Node.js 18 on Debian Buster on ARM64
+    needs: create-release
+    runs-on: ubuntu-latest
+    continue-on-error: true
+
+    steps:
+    - uses: actions/checkout@v3
+      with:
+        submodules: true
+    - name: Set up QEMU
+      uses: docker/setup-qemu-action@v3
+      with:
+        platforms: arm64
+    - name: Install, test, and create artifact
+      uses: ./.github/actions/linux-arm64-node-18/
+      env:
+        GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}

